### ISSUE TITLE
Getting Visual Studio's Live Unit Testing to work?

### ISSUE THREAD

**Issue by aybe** [Rabadash8820/UnityAssemblies#17]
Here's a quick guide on how to get Live Unit Testing to work when using Unity assemblies.

First step is to decide of a folder in `Assets` that is to become testable, e.g. `Assets/MyProject`.

- in the solution, add a solution folder `Tests`
- in that folder, create a NET Standard 2.1 class library in like `Projects/MyProject`
- setup its `Directory.build.props` as follows:
```
<Project>
	<PropertyGroup>
		<UnityProjectPath>$(SolutionDir)</UnityProjectPath>
	</PropertyGroup>
</Project>
```
- setup its `.csproj` as follows:
```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>netstandard2.1</TargetFramework>
		<Nullable>enable</Nullable>
		<AllowUnsafeBlocks>true</AllowUnsafeBlocks>
		<LangVersion>9</LangVersion>
	</PropertyGroup>

	<!--enable assertions-->
	<PropertyGroup>
		<DefineConstants>$(DefineConstants);UNITY_ASSERTIONS</DefineConstants>
	</PropertyGroup>

	<ItemGroup>

		<PackageReference Include="Unity3D" Version="2.1.3" />
		
		<!--fix incoming ambiguous references-->
		<Reference Remove="$(UnityEnginePath)" />

		<!--enable assertions-->
		<Reference Include="$(UnityModulesPath)\UnityEngine.CoreModule.dll" Private="True" />

		<!--enable extra stuff-->
		<Reference Include="$(UnityScriptAssembliesPath)\Unity.Mathematics.dll" Private="True" />

		<!--setup project content-->
		<Compile Include="$(SolutionDir)Assets\MyProject\**\*.cs" />

	</ItemGroup>

</Project>
```
- reference that new project into your MSTest project
- things should now work as they usually do for LUT

There are two tricks involved here:
 - referencing files as links, project will sync itself with content in `Assets/MyProject`
 - `Private="True"` is what makes LUT to work

If you still run into trouble, try the following steps:
- close Visual Studio
- remove `C:\Users\USERNAME\source\lut` folder, e.g. `rd lut /s /q`
  - that projected file system sometimes cannot be deleted, just keep trying
- remove `.vs` folder in the folder project
- remove `.lutconfig` and create a new one from _Test/Live Unit Testing/Configure..._ menu
  - for exclusion, pick _Common artifacts eg. bin/, obj/, *.dll, *.obj_
- start LUT again, it should work as expected

**Extra notes:**

When you add new files, do so in `Assets/MyProject`, switch to/from Unity so that VS updates/reloads the solution.

`.gitignore` can be modified like so:

```
TODO drop here https://github.com/github/gitignore/blob/main/VisualStudio.gitignore

TODO drop here https://github.com/github/gitignore/blob/main/Unity.gitignore

# keep project so it can be referenced by MSTest project (fix Unity.gitignore)
!/**/*.csproj

# but ignore Unity-generated projects at root
/*.csproj

# keep meta for Unity (fix VisualStudio.gitignore)
!/Assets/**/*.meta

# keep Unity packages (fix VisualStudio.gitignore)
!**/[Pp]ackages/*
```

The result is amazing:
- use Unity without hardcoding references
- project is immediately usable within Unity
- project still works with Live Unit Testing

🥳

**Update:**

Check [this](https://github.com/Rabadash8820/UnityAssemblies/issues/17#issuecomment-1716967007) to fix _ECall methods must be packaged into a system module_.

**Comment by aybe** [comment]
Editing this post for clarity, solution is [here](https://github.com/Rabadash8820/UnityAssemblies/issues/17#issue-1869007454).

**Comment by aybe** [comment]
Editing this post for clarity, solution is [here](https://github.com/Rabadash8820/UnityAssemblies/issues/17#issue-1869007454).

**Comment by Rabadash8820** [comment]
@aybe Glad you were able to find a setup that works! I'm not super familiar with LUT, but you might wanna read [Referencing assemblies stored in a Unity project](https://github.com/Rabadash8820/UnityAssemblies#referencing-assemblies-stored-in-a-unity-project) in the Readme if you haven't already. It goes into detail about setting up this NuGet when your solution and Unity project are in the same repo. You can see the test solution and Unity projects in this repo for examples.

**Comment by Rabadash8820** [comment]
Oops, didn't mean to reopen this ticket. 😛

**Comment by aybe** [comment]
Editing this post for clarity, solution is [here](https://github.com/Rabadash8820/UnityAssemblies/issues/17#issue-1869007454).

**Comment by Rabadash8820** [comment]
Glad that helped!

Tbh, running tests on every save sounds much more reasonable then running every single time I type a character...the vast majority of those runs would fail due to compiler errors anyway, no?  Constantly running hundreds of unit tests on any sizable project also sounds like a nightmare for performance. Even running on every save can be annoying; I used to run Jest in watch mode for JS code, but I save constantly, so it led to constant waiting for the entire test suite to finish before I could see results for the specific test that I was editing. I much prefer to just trigger specific tests manually after completing a change, and then run them all before I commit. That said, I know a lot of people swear by LUT, so I'll have to read up on it more!

**Comment by Rabadash8820** [comment]
On a side note, if you weren't aware, you can edit GitHub comments. Sometimes that makes it easier to update a "recommended solution" like what you're showing in this Issue.

**Comment by aybe** [comment]
Yep, you're right, it's a bad habit I got from online forums 🤣, edited the topic so it points to a single place!

---

Well of course you cannot/should not run LUT on all of the tests but only a selection.

I can only recommend to [read the docs](https://learn.microsoft.com/en-us/visualstudio/test/live-unit-testing?view=vs-2022), it's explained very clearly.

The different icons shown in each LOC in real-time is what makes it truly insane.

Practical example:

I had to do FIR convolution from [OnAudioFilterRead](https://docs.unity3d.com/ScriptReference/MonoBehaviour.html), did it in managed code but performance was terrible.
No matter how hard I optimized it, the VU meter always stayed red .

Went Unity Burst but round-trip was pretty expensive with profiling every time.
So I wrote a couple of algorithms to full/half-band, vectorization types, leverage taps symmetry.

To get these rights, unit tests are a must, where LUT came in handy is that I'd immediately see results.
Getting the right indices in delay line for some proved **much** trickier than I initially thought. 🫠

Now imagine how many times I'd click _Run tests_ manually, 1000s of times, but with LUT, zero!

That's pretty much the idea, truly simple but so effective.

At some point, I thought that since iterating in this environment is much faster, I should get them Unity-ready.
Basically, go pointer route so that the code becomes `[BurstCompile]` valid. Done today, time for fun now!

This is my workspace, 12 algos, 20 filters, 240 tests to cover many cases, immediate feedback:

<img width="536" alt="image" src="https://github.com/Rabadash8820/UnityAssemblies/assets/1537591/85584d02-be50-4ca8-8edb-e6189a8aa47e">

And thanks to your project 🎯💯, I got the Burst-ready step ready in no time!

**Comment by Rabadash8820** [comment]
Nice! That is a pretty sweet looking setup. You've convinced me to check it out at least lol. Good luck on this intense audio project!

**Comment by aybe** [comment]
Thank you and thanks for this great project :)

**Comment by aybe** [comment]
Came across an interesting issue: solving the dreaded _"ECall methods must be packaged into a system module."_! 🤪

Basically, as soon as I added profiler markers to the Unity project, the code would stop being testable because of this.

Microsoft Fakes was not able to solve that as it doesn't support structs and external methods...

So I thought, why not try to lure the system with `global using` once again?

Turns out that it worked and is much better since original code needs zero modifications! 🥳

The key to solve this, along `global using`, was to put the following in the decoy class library project:

```
global using ProfilerMarker = Unity.Profiling.Fakes.ShimProfilerMarker;
using System;
using System.Diagnostics.CodeAnalysis;

// ReSharper disable CheckNamespace

namespace Unity.Profiling.Fakes;

internal readonly struct ShimProfilerMarker
{
    [SuppressMessage("ReSharper", "UnusedParameter.Local")]
    [SuppressMessage("Style", "IDE0060:Remove unused parameter")]
    public ShimProfilerMarker(ProfilerCategory category, string name)
    {
    }

    [SuppressMessage("ReSharper", "MemberCanBeMadeStatic.Global")]
    public AutoScope Auto()
    {
        return new AutoScope();
    }

    public readonly struct AutoScope : IDisposable
    {
        public void Dispose()
        {
        }
    }
}
```

**Comment by aybe** [comment]
Hi Dan,

My project is up and I've credited you because yours was of great help to get it up. :)

Thanks!

https://github.com/aybe/FIRConvolution

**Comment by Rabadash8820** [comment]
Hey, thank you so much! Hope your project does well!

**Comment by aybe** [comment]
Thank you mate, take care! :)

**Comment by aybe** [comment]
Quick update on how to do it with V3:

General notes:

- ```Private``` must be ```True``` as opposed to docs, else LUT will complain that it doesn't find references
- make sure to to add ```Library/``` to ```.lutignore``` to avoid copying everything to LUT scratch folder
- since ```$(UnityScriptAssembliesPath)``` points to ```\Library\ScriptAssemblies```, MSB3106 will occur
  - the project below sits in ```C:\MyUnityGame\Projects\MyUnitTest\MyUnitTest.csproj```
  - to address MSB3106, prefix ```$(UnityScriptAssembliesPath)``` with ```$(SolutionDir.TrimEnd('\'))```

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>net8.0</TargetFramework>
		<LangVersion>latest</LangVersion>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.14.1" />
		<PackageReference Include="MSTest" Version="3.11.0" />
	</ItemGroup>

	<ItemGroup>
		<Using Include="Microsoft.VisualStudio.TestTools.UnitTesting" />
	</ItemGroup>

	<ItemGroup>
		<ProjectReference Include="..\..\Wipeout.Engine.UI.csproj" />
	</ItemGroup>

	<PropertyGroup>
		<UnityVersion>6000.2.6f1</UnityVersion>
		<IncludeDefaultUnityAssemblyReferences>False</IncludeDefaultUnityAssemblyReferences>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Unity3D" Version="3.0.1" />
		<Reference Include="$(UnityModulesPath)\UnityEngine.CoreModule.dll" Private="True" />
		<Reference Include="$(SolutionDir.TrimEnd('\'))$(UnityScriptAssembliesPath)\Unity.Burst.dll" Private="True" />
	</ItemGroup>

</Project>
```

EDIT: MSB3106 fixed, procedure updated
